### ISSUE TITLE
Confirming workflow to create dataloader / dataset from mIF slides

### ISSUE THREAD

**Issue by surya-narayanan** [Dana-Farber-AIOS/pathml#218]
Hey, I wanted to confirm my workflow, to ensure that I’m doing it right. I want to create a data loader of tiles from mIF slides, without running any pre-processing on the slides.

Since a tiledataset takes a file path, should I then save the slide to a h5 file first? That seems wasteful.  

I am confused why ‘_generate_tiles_’ is a property of the backend and not of the slide object. Ideally, I want a function that takes a SlideDataset and returns a TileDataset. 

I take it there's no way to directly go from a SlideDataset to a TileDataset? 

Thanks.

**Comment by jacobrosenthaldfci** [comment]
Hi Surya, so the way this would fit into the workflow as we have it now is to run a pass-through pipeline and write it to h5path, then create the TileDataset from that.

Pseudocode:
````python
dataset = SlideDataset([<list here>])
pipe = Pipeline()
dataset.run(pipe, distributed=False)
dataset.write("/path/to/write/")
tile_dataset = TileDataset("/path/to/write/")
````

It may seem unnecessary to write a h5 file first, but the h5 format is optimized for dataloaders, so it's probably faster this way than to create a dataset object which reads tiles on the fly from the raw images. For example, reading a tile through the bioformats backend would be much much slower than reading a tile which has already been pulled out and put into h5, because it has to start a java VM, open the file, etc. etc. vs just looking at the index of the tile and pulling it directly from the h5 file. So there's a tradeoff between disk space and performance and we chose to err on the side of using more disk space for hopefully faster performance. Hope this is helpful

**Comment by surya-narayanan** [comment]
Cool. That's helpful. Is the disk space required 100% of the original? So if I have 1 TB of WSIs, will it write 1 TB of tiles? Thanks again.

**Comment by jacobrosenthaldfci** [comment]
I would guess that disk space usage would be similar, maybe slightly more (depending on if or how much the original image is compressed) but probably same order of magnitude.

If you are not processing the tiles, I don't expect to see much (any?) speedup from distributed processing, in fact the overhead of spinning up the cluster and sending data to and from the workers may end up making it slower in this context. I have not benchmarked this, but that's my intuition

**Comment by surya-narayanan** [comment]
Ok. But this seems like it could take 25 + hours for a slide dataset of 90 slides (which is roughly 1/3 of my dataset). Are you sure that there's nothing I can do to speed this up?

My workflow is exactly as you mentioned previously - 


```
#Get paths to slides
slide_paths = [glob command that spits out path to files]

#Create a pipeline
pipeline = Pipeline([])

#Create slide objects
slides = [SlideData(i, backend = "bioformats", slide_type = types.Vectra) for i in slide_paths]

#Run pipeline on each slide
for slide in tqdm(slides): #Could create slide dataset object here but manually highjacking to measure time performance, since single-threaded and slidedataset.run does exactly the same thing

  slide.run(pipeline, distributed=False)
  slide.write(f'/mnt/disks/ip_tiles/ip_slides_pathml/{slide.name}.h5')
```

**Comment by jacobrosenthaldfci** [comment]
We're working on performance improvements, you can track our progress at https://github.com/Dana-Farber-AIOS/pathml/issues/225

**Comment by surya-narayanan** [comment]
Hi, 

I was able to parallelize this operation and reduce the runtime by leveraging multiple workers. I understand you use Dask to manage the parallelization, but here's what I did - 

```
def run_and_write(slide_name):
  slide = SlideData(slide_name, backend = "bioformats", slide_type = types.Vectra)
  pipeline = Pipeline([])
  slide.run(pipeline, distributed=False)
  slide.write(f'/mnt/disks/ip_tiles/ip_slides_pathml/{slide.name}.h5')

p_umap(run_and_write, slides_names)

```

Where `p_umap` is a multi-processing function that distributes the function `run_and_write` and the members of the `slides_names` list to different workers. So my speedup was proportional to the number of workers I had. 

I'm curious on your thoughts on this.
